feat: experimental skip domain injection

AtofStryker opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes #24290
- Closes #23958
- Closes #2367 (with `experimentalModifyObstructiveThirdPartyCode` option enabled)
- Closes #24418

### User facing changelog
Adds the `experimentalSkipDomainInjection ` option to disable Cypress from setting `document.domain` on injection. 

The `experimentalSkipDomainInjection ` option is either an array of strings/globs to match against url hostnames. If there is a match, `document.domain` will not be injected. This means that all cross-origin navigations in the provided config, including sub domain navigations, will require a `cy.origin()` block for Cypress to interact with the page.

### Additional details
The goal of this flag is to unblock Cypress users who are running into `document.domain` issues with the sites they are testing, which includes Salesforce Winter '22. This can also be a diagnostic tool in the future as well for other sites where we expect `document.domain` to possibly be an issue, and we can simply ask users to toggle on the flag and see if the situation improves

Currently the goal of this experimental flag is to unblock users with `document.domain`. We want to observe this experiment usage which should educate how it evolves over time. There are no plans to make this feature GA, and at best this may be a supported configuration option in the future that is not experimental (unless setting document.domain is fully removed, in which case we may revisit this).

`cypress-documentation` likely will need to add info snippets in the following areas, including information on the experimental flag (added in https://github.com/cypress-io/cypress-documentation/pull/4955):
* https://docs.cypress.io/guides/guides/web-security#Examples-of-what-Cypress-does-under-the-hood
* https://docs.cypress.io/guides/guides/cross-origin-testing#Parts-of-a-URL

### Steps to test
Since this change requires a experimental config option that the server needs to be aware of, system tests were added to fully test the expected behavior in regards to navigation

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?https://github.com/cypress-io/cypress-documentation/pull/4955
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

AtofStryker committed:
feat: set up experimentalUseDefaultDocumentDomain to disallow document.domain overwritting

AtofStryker committed:
use default domain around experimentalUseDefaultDocumentDomain in main iframe and spec bridge iframes. Also adapt CORS policy to use same-origin if experimental flag is set

AtofStryker committed:
run ci

AtofStryker committed:
fix: add insertion of experimental flag where is was needed/missing

AtofStryker committed:
chore: add system test to exercise experimental flag for expected behavior

AtofStryker committed:
fix: fix issues with template updates to conform to squirrelly v7

AtofStryker committed:
fix: update config tests to include new experimental flag

AtofStryker committed:
run ci

AtofStryker committed:
fix: trailing whitespace [run ci]

AtofStryker committed:
chore: update snapshot

AtofStryker committed:
run ci

AtofStryker committed:
Merge branch 'develop' into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
fix: update proxy unit tests to account for experimentalUseDefaultDocumentDomain

AtofStryker committed:
run ci

flotwig commented on cli/types/cypress.d.ts:
```diff
@@ -3014,6 +3014,15 @@ declare namespace Cypress {
+     * @see https://developer.mozilla.org/en-US/docs/Web/API/Document/domain
+     * @default false
+     */
+    experimentalUseDefaultDocumentDomain: boolean
```
What changes would it take to make this the default behavior for all users (taking it out of experimental)?

AtofStryker replied:
I think it depends where we land, whether its an "all or nothing" approach like what's implemented right now or something that is configurable per domain. As for now, the only thing we would need to do to take this out of experimental is making the `policyForDomain` always `same-origin` and reverse the conditional which is pretty straight forward. And the change would be breaking.

However I think right now it likely may not work with same origin iframes within the AUT, since we need partial `document.domain` injection for same-origin `localhost` iframes. There might need to be some caveats around this, which would probably be a good idea to document.

mike-plummer committed:
fix: Allow component tests with special characters in filepath (#25299)

feat: cut over experimental flag to take list of known problematic domains via string/glob pattern

run ci

chore: update system test and fix broken config

AtofStryker committed:
fix: fix server unit and integration tests. integration tests should no longer use google to test against injection as we do not inject document.domain on google domains

AtofStryker committed:
run ci

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
run ci

AtofStryker committed:
fix: server integration tests where google documents are expected to receive document.domain injection. Kept test same by changing URL

AtofStryker committed:
run ci

AtofStryker committed:
fix: update server test with mssing unupdated assertions

AtofStryker committed:
run ci

AtofStryker committed:
fix: turn off experimental flag by default while recommending sane defaults to users to configure

AtofStryker committed:
run ci

AtofStryker committed:
chore: fix typings [run ci]

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
run ci

AtofStryker committed:
chore: make experiment an e2e option only

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
run ci

AtofStryker committed:
Merge branch 'develop' into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
Merge branch 'develop' into feat/experimentalDefaultDocumentDomain

chrisbreiding commented on packages/proxy/lib/http/request-middleware.ts:
```diff
@@ -162,8 +162,8 @@ const MaybeEndRequestWithBufferedResponse: RequestMiddleware = function () {
-    // NOTE: Only inject fullCrossOrigin here if experimental is on and
-    // the super domain origins do not match in order to keep parity with cypress application reloads
+
+    // NOTE: Only inject fullCrossOrigin here the super domain origins do not match in order to keep parity with cypress application reloads
```
```suggestion
    // NOTE: Only inject fullCrossOrigin here if the super domain origins do not match in order to keep parity with cypress application reloads
```

AtofStryker replied:
addressed in dccf967ed9

emilyrohrbough commented on cli/types/cypress.d.ts:
```diff
@@ -3019,6 +3019,17 @@ declare namespace Cypress {
      */
     experimentalModifyObstructiveThirdPartyCode: boolean
+    /**
+     * Disables setting document.domain to the document's super domain on injection.
```
documents or applications super domain?

AtofStryker replied:
I think in this case they are the same thing, but application might be easier to digest?

AtofStryker replied:
renamed in 86bea96010  (meant to do in dccf967ed9)

emilyrohrbough commented on cli/types/cypress.d.ts:
```diff
@@ -3019,6 +3019,17 @@ declare namespace Cypress {
+    /**
+     * Disables setting document.domain to the document's super domain on injection.
+     * This experiment is to be used for sites that do not work with setting document.domain
+     * due to cross-origin issues. Enabling this option no longer allows for default sub domain
```
```suggestion
     * due to cross-origin issues. Enabling this option no longer allows for default subdomain
```

AtofStryker replied:
addressed in dccf967ed9

emilyrohrbough commented on cli/types/cypress.d.ts:
```diff
@@ -3019,6 +3019,17 @@ declare namespace Cypress {
+     * This experiment is to be used for sites that do not work with setting document.domain
+     * due to cross-origin issues. Enabling this option no longer allows for default sub domain
+     * navigations, and will require the use of cy.origin(). This option takes an array of
+     * strings/string globs, and applies defaults for known domains where document.domain is known
```
> and applies defaults for known domains where document.domain is known

I think we should drop this portion. I assu,e you mean google, but that's happening today without this enabled.

AtofStryker replied:
I think I originally assumed the defaults we applied but that context makes sense. removed in dccf967ed9

emilyrohrbough commented on packages/config/src/options.ts:
```diff
@@ -215,6 +215,12 @@ const driverConfigOptions: Array<DriverConfigOption> = [
+  }, {
+    name: 'experimentalUseDefaultDocumentDomain',
+    defaultValue: null,
+    validation: (_: string, value: any) => validate.isArrayOfStrings(value) || value === null,
```
We need to enhance this a bit, if value is not an array of strings and value is not null, we wont' have a meaningful error message / will throw an error from returning an invalid error response for the error logging in the launchpad/reporter.

If `value === null` we need to return these parts:
```js
  return {
    key: 'experimentalUseDefaultDocumentDomain'
    value,
    type: 'an array of strings or null'
  }
```

AtofStryker replied:
fixed in dccf967ed9

emilyrohrbough commented on packages/config/src/validation.ts:
```diff
@@ -33,7 +33,7 @@ const _isFullyQualifiedUrl = (value: any): ErrResult | boolean => {
 
-const isArrayOfStrings = (value: any): ErrResult | boolean => {
+export const isArrayOfStrings = (value: any): ErrResult | boolean => {
   return _.isArray(value) && _.every(value, _.isString)
```
Same comment as above, this isn't currently wired up to return the correct error response for logging if this fails, it needs to uses to return `errMsg` if this condition fails. You'll need to check to ensure this doesn't impact `isStringOrArrayOfStrings`.

AtofStryker replied:
Ah that's starting to make a lot more sense! I think what I might do then to lessen the impact is just create a `isNullOrArrayOfStrings` that returns `errMsg` or `true` similar to `isStringOrArrayOfStrings`. That way things are at least way more consistent with what we have.

AtofStryker replied:
fixed in dccf967ed9

emilyrohrbough commented on packages/driver/src/cypress.ts:
```diff
@@ -182,7 +183,11 @@ class $Cypress {
     // set domainName but allow us to turn
     // off this feature in testing
-    if (domainName && config.testingType === 'e2e') {
+    const shouldInjectDocumentDomain = cors.shouldInjectDocumentDomain(window.location.origin, {
```
Isn't this false when `experimentalUseDefaultDocumentDomain` is true? I'm hesitant to pull this module into the driver code. Most of the cors code is expected to run in node not the browser right?

emilyrohrbough replied:
I think it would be better to update the logic in the data-context/server to no sent the `domainName` name if this should not be injected by default into a new Cypress instance.

AtofStryker replied:
> Isn't this false when experimentalUseDefaultDocumentDomain is true?

the other way around. if `experimentalUseDefaultDocumentDomain` is truthy and fits the glob patterns, then `shouldInjectDocumentDomain` should return false.

The cors code is already referenced in the driver quite a bit. I think the expectation is that is should run in the browser or node, but not sure if that was the original intention 😅 .

emilyrohrbough replied:
@AtofStryker is it the driver or just the driver tests? I wasn't finding the dependency used in the shipped code when i searched for it, but I must be missing it

AtofStryker replied:
I'm hesitating to modify this in the data-context and server for the manipulated domainName value because it isn't exactly clear why that value is being manipulated since it would happen in the driver. The config flag in the `configure` method is localized, which I think makes the impact/expected change clearer.

AtofStryker replied:
noticed I had a bug in the cors package around the `shouldInjectDocumentDomain` function where it wasn't returning false on glob matches. going to add some unit tests here

AtofStryker replied:
fixed the bug in 861de35229

emilyrohrbough commented on packages/network/package.json:
```diff
@@ -20,6 +20,7 @@
     "debug": "^4.3.2",
     "fs-extra": "9.1.0",
     "lodash": "^4.17.21",
+    "minimatch": "^3.0.5",
```
Should we keep this `3.0.5` to match the versioning of minimatch used throughout the repo?

AtofStryker replied:
That was the intent. I kept it at `^3.0.0` for that reason for consistency of the API. We can also upgrade it but might be confusing when comparing against different packages

emilyrohrbough replied:
The others are locked down though 😬 
```suggestion
    "minimatch": "3.0.5",
```

AtofStryker replied:
ill pin it!

emilyrohrbough commented on packages/server/lib/controllers/files.js:
```diff
@@ -26,9 +26,15 @@ module.exports = {
 
       debug('all files to send %o', _.map(allFilesToSend, 'relative'))
 
+      const documentSuperDomainIfExists = cors.shouldInjectDocumentDomain(req.proxiedUrl, {
```
Can we align this throughout as `shouldInjectDocumentDomain`. I think this is clearer than `documentSuperDomainIfExists`
```suggestion
      const shouldInjectDocumentDomain = cors.shouldInjectDocumentDomain(req.proxiedUrl, {
```

AtofStryker replied:
The thing is this value is actually the domain value if it exists. the `shouldInjectDocumentDomain` is just the gate for it. `documentSuperDomainIfExists` in this case is either going to be the super domain string, or undefined. `shouldInjectDocumentDomain` makes the variable read like a boolean when it isn't

emilyrohrbough replied:
I def read this as a boolean because of the `ifExists`. maybe `superDomain` is sufficent and if it has a value in the template it'll read like
```
if(superdomain) {
  document.domain = superdomain
}
```

AtofStryker replied:
awesome! Addressed in 9b6938f5be

emilyrohrbough commented on packages/server/lib/html/spec-bridge-iframe.html:
```diff
@@ -6,7 +6,9 @@
   <body>
     <script type="text/javascript">
-      document.domain = '{{domain}}';
+  {{if(options.documentSuperDomainIfExists)}}
```
should this be four spaces?

AtofStryker replied:
trying to get these [squirrelly](https://www.npmjs.com/package/squirrelly) templates to format correctly on injection isa bit of a pain. It looks ugly here, but formats correctly on injection. I can double check and also see if there is room to clean this up a bit!

AtofStryker replied:
The template looks ugly but that seems to be the way to get it displayed correctly? 
![](https://user-images.githubusercontent.com/3980464/210620414-41fb8aa5-89e2-4af2-a3e8-b0b2e728ce05.png)

The [docs](https://v7--squirrellyjs.netlify.app/docs/v7/cheatsheet#conditionals) didn't really explain it but the template also seems to add spacing?

emilyrohrbough commented on packages/server/lib/server-e2e.ts:
```diff
@@ -45,6 +45,8 @@ const isResponseHtml = function (contentType, responseBuffer) {
 export class ServerE2E extends ServerBase<SocketE2E> {
   private _urlResolver: Bluebird<Record<string, any>> | null
+  // the initialization of this variable is only precautionary as the actual config value is applied when the server is created
+  private useDefaultDocumentForDomains: string[] | null = null
```
Reading through this, I really want this to be something like `excludeDomainInjectionPatterns` or `excludeDomainInjectionList`. I feel like I really only understand `useDefaultDocumentForDomains` because I'm currently reviewing the concept.

emilyrohrbough replied:
I also feel like I want this to carry over into the experimental Name. Like `experimentalExcludeDomainInjectionPattern`  or `excludeDocumentDomainList`. I don't even love these names 🤷🏻‍♀️

AtofStryker replied:
this 💯 . I don't think `experimentalUseDefaultDocumentDomain` is a good name as it isn't entirely representative of what is happening, and it reads like a boolean when the options are a string array or null

AtofStryker replied:
renamed `experimentalUseDefaultDocumentDomain` and the `useDefaultForDomains` naming to using the `experimentalSkipDomainInjection` nomenclature in 86bea96010

emilyrohrbough reviewed (commented):
I would have expected a new error message or two for cross-origin errors for when this is set to remind users they excluded this domain and cy.origin() is required for this specific super domains.

AtofStryker commented:
> I would have expected a new error message or two for cross-origin errors for when this is set to remind users they excluded this domain and cy.origin() is required for this specific super domains.

Currently we are just reusing the `same-origin` / `same-super-domain` messaging in [error_messages](https://github.com/cypress-io/cypress/blob/feat/experimentalDefaultDocumentDomain/packages/driver/src/cypress/error_messages.ts#L1195), which for the domain configured would error for an origin mismatch and not domain. But right now I think Google is the only exception to this rule, so I wonder if we can add a plug for the experiment and if the flag is turned on say something like " Since experiment is on and you have this domain configured, you'd need a cy.origin() block when performing cross origin navigations". Do you think something like that would be helpful? Right now I feel the error is a  bit contextless

AtofStryker committed:
chore: address comments in code review

AtofStryker committed:
chore: rename experimentalUseDefaultDocumentDomain to experimentalSkipDomainInjection

AtofStryker committed:
Merge branch 'feat/experimentalDefaultDocumentDomain' of github.com:cypress-io/cypress into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
fix regression in shouldInjectionDocumentDomain utility function and add unit tests

AtofStryker commented:
@emilyrohrbough the cross origin error looks something like this for subdomains with the option configured correctly:
![](https://user-images.githubusercontent.com/3980464/210618988-f1c6e7d3-338e-4dac-aa5a-d3c53b7856be.png)

AtofStryker committed:
run ci

AtofStryker commented on packages/errors/src/errors.ts:
```diff
@@ -1185,6 +1185,19 @@ export const AllCypressErrors = {
+    const code = errPartial`
+    {
+      e2e: {
+        experimentalSkipDomainInjection: ['*.salesforce.com', '*.force.com', '*.google.com', 'google.com']
```
should I go with something less specific here?

alexanderg1982 replied:
I've opened the #24418 issue, and I think this should be changed for the following domains also: *.my.site.com 
taken from https://help.salesforce.com/s/articleView?id=sf.domain_name_enhanced.htm&type=5

AtofStryker replied:
@alexanderg1982 this should resolve #24418. This line here is more so to recommend users on how to configure the option if the config isn't inside the `e2e` object. I think it's too specific since we just want to recommend where to move the option. When we release this, you should be able to do:

```js
experimentalSkipDomainInjection: ['power-inspiration-1088-dev-ed.scratch.my.site.com']
```

emilyrohrbough replied:
@AtofStryker Maybe it would be good to call out this is a recommendation?

AtofStryker replied:
I'm wondering if we just want to reference the enhanced domain names in the docs?

AtofStryker replied:
@alexanderg1982 I added the enhanced domain reference in the [docs](https://github.com/cypress-io/cypress-documentation/pull/4955/files#diff-8df10062ef7f8e59d80d177d29301414b5c83d2fb7c52681870439be2de71b80R97) to help guide the correct domains. Thank you for sharing this!

AtofStryker committed:
chore: rename documentSuperDomainIfExists to superDomain [run ci]

emilyrohrbough commented on packages/network/lib/cors.ts:
```diff
@@ -158,15 +161,53 @@ export const urlSameSiteMatch = (frameUrl: string, topUrl: string): boolean => {
 }
 
+/**
+ *
```
or add description 
```suggestion
```

AtofStryker replied:
addressed in e27acf72e2

emilyrohrbough commented on packages/network/lib/cors.ts:
```diff
@@ -158,15 +161,53 @@ export const urlSameSiteMatch = (frameUrl: string, topUrl: string): boolean => {
 /**
  * Returns the policy that will be used for the specified url.
  * @param url - the url to check the policy against.
  * @returns a Policy string.
```
update to add new opts param

AtofStryker replied:
addressed in e27acf72e2

emilyrohrbough commented on packages/network/lib/cors.ts:
```diff
@@ -177,9 +218,13 @@ export const policyForDomain = (url: string): Policy => {
  * @param topUrl - The url you are testing the policy in context of.
  * @returns boolean, true if matching, false if not.
```
add new param to description

AtofStryker replied:
addressed in e27acf72e2

emilyrohrbough commented on packages/network/lib/cors.ts:
```diff
@@ -193,9 +238,12 @@ export const urlMatchesPolicyBasedOnDomain = (frameUrl: string, topUrl: string):
+export const urlMatchesPolicyBasedOnDomainProps = (frameUrl: string, topProps: ParsedHostWithProtocolAndHost, opts?: {
+  skipDomainInjectionForDomains: string[]
+}): boolean => {
+  const policy = policyForDomain(frameUrl, {
```
Policy for domain check if `skipDomainInjectionForDomains` is an array. this should just pass through instead of providing a fallback.
```suggestion
  const policy = policyForDomain(frameUrl, opts)
```

AtofStryker replied:
addressed in 8c6813dbf3

emilyrohrbough commented:
> " Since experiment is on and you have this domain configured, you'd need a cy.origin() block when performing cross origin navigations". Do you think something like that would be helpful? Right now I feel the error is a bit contextless

I do think this would be helpful, especially because there are workflows working for users when on Firefox so it'd could be helpful to provide more direction.

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
chore: address comments from code review

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into feat/experimentalDefaultDocumentDomain

AtofStryker committed:
chore: just pass opts through to policyForDomain

AtofStryker commented:
Added some context in the cross origin error @emilyrohrbough in e27acf72e2fd8f636fb0a371b61af4b317313726 . It isn't smart enough right now to compare the url, but we could add that in the future or if we really think it would be helpful right now

AtofStryker committed:
run ci
